server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.action.admin.cluster.node.shutdown.NodesRemovalPrevalidation.IsSafe;
+import static org.elasticsearch.action.admin.cluster.node.shutdown.NodesRemovalPrevalidation.NodeResult;
+import static org.elasticsearch.action.admin.cluster.node.shutdown.NodesRemovalPrevalidation.Result;
+
+public class TransportPrevalidateNodeRemovalAction extends TransportMasterNodeReadAction<
+    PrevalidateNodeRemovalRequest,
+    PrevalidateNodeRemovalResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportPrevalidateNodeRemovalAction.class);
+
+    @Inject
+    public TransportPrevalidateNodeRemovalAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            PrevalidateNodeRemovalAction.NAME,
+            false,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            PrevalidateNodeRemovalRequest::new,
+            indexNameExpressionResolver,
+            PrevalidateNodeRemovalResponse::new,
+            ThreadPool.Names.SAME
+        );
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        PrevalidateNodeRemovalRequest request,
+        ClusterState state,
+        ActionListener<PrevalidateNodeRemovalResponse> listener
+    ) {
+        List<String> nodes = request.getNodeIds();
+        try {
+            Set<DiscoveryNode> concreteNodes = resolveNodes(nodes, state.nodes());
+            request.setConcreteNodes(concreteNodes.toArray(new DiscoveryNode[0]));
+            doPrevalidation(request, state, listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    public static Set<DiscoveryNode> resolveNodes(List<String> nodes, DiscoveryNodes discoveryNodes) {
+        Set<DiscoveryNode> concreteNodes = new HashSet<>(nodes.size());
+        for (String node : nodes) {
+            List<DiscoveryNode> matches = discoveryNodes.stream()
+                .filter(dn -> dn.getId().equals(node) || dn.getName().equals(node) || dn.getExternalId().equals(node))
+                .toList();
+            if (matches.isEmpty()) {
+                throw new ResourceNotFoundException("node [{}] not found", node);
+            }
+            if (matches.size() > 1) {
+                throw new IllegalArgumentException("more than one node matches [" + node + "]");
+            }
+            concreteNodes.add(matches.get(0));
+        }
+        if (concreteNodes.size() != nodes.size()) {
+            throw new IllegalArgumentException(
+                Strings.format("provided {} values for <nodes> which resolved to {} nodes", nodes.size(), concreteNodes.size())
+            );
+        }
+        ;
```

Comment:
I have tried to add some validations here to make sure prevalidation calls do not accidentally lead to a false safe removal response. By that I mean, since we do provide an overall result, if the provided nodes (name or id) do not each uniquely identify a node, I do not consider the request valid. I am assuming it is not safe to dismiss an id (perhaps with a typo) if there is no matching node for it. I am not sure whether this is strictly necessary, so I'd appreciate a second opinion.

Reply:
I think the ambiguous case (more than one node matches) is fine to throw on. I think I'd like to provide a response also if we are missing a node - but the response should indicate that it was missing. That way we can leave it up to the orchestration to make a decision - that might depend on the situation (vacate/system vs user initiated change).